[ConstraintSystem] Diagnose missing AnyObject conformance using the MissingConformance constraint fix.

hborla opened the pull request:
Additionally, this change adds a special note for requirement failures where the type that failed the requirement is `Optional`, but the `Wrapped` type satisfies the requirement. This is done instead of using the `ForceOptional` constraint fix for two reasons:
1. The locator for a failed requirement points to the generic requirement. We don't know what caused us to bind a particular type to a type variable, so it's tricky to figure out which subexpression should be unwrapped, or if there is one.
2. There isn't always a subexpression to unwrap, but it's still useful to know that the wrapped type satisfies the requirement. For example:
```swift
class C {}
func genericFunc<T: AnyObject>(_ t: [T]) {} // note: where 'T' = 'C?'

let arr: [C?] = []
genericFunc(arr) // error: global function 'genericFunc' requires that 'C?' conform to 'AnyObject'`
// note: wrapped type 'C' satisfies this requirement
```

hborla committed:
[ConstraintSystem] Diagnose missing AnyObject conformance using
the MissingConformance constraint fix.

hborla committed:
[Diagnostics] Add a tailored note for when the `Wrapped` type of an
optional satisfies a failed requirement.

hborla commented:
@swift-ci please smoke test

hborla commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -417,9 +417,10 @@ class MissingConformanceFailure final : public RequirementFailure {
-      : RequirementFailure(cs, expr, RequirementKind::Conformance,
-                           conformance.first, conformance.second, locator) {}
+                            std::pair<Type, Type> conformance,
+                            RequirementKind kind)
```
In `RequirementFailure` constructor, we pass in the `RequirementKind` and check:
```cpp
    auto reqElt = locator->castLastElementTo<LocatorPathElt::AnyRequirement>();
    assert(reqElt.getRequirementKind() == kind);
```
If we're going to get `kind` from the locator anyway, is this necessary? `kind` isn't used anywhere else in the constructor. I think it makes more sense to perform assertions when constructing the various subclasses of `RequirementFailure`. What do you think?

xedin replied:
Yes, that's why I mentioned that locator has `kind` now, I think we can just use that instead of passing it as an argument. Originally `TypeParameterRequirement` didn't have any information, but once I have extended it first with index then kind I didn't update all of the places :/

hborla committed:
[ConstraintFix] Don't pass along the RequirementKind to the MissingConformance
constraint fix. Instead, get the kind from the locator.

jrose-apple commented:
You don't have to do this yet, but "conform to 'AnyObject'" feels misleading, especially in today's Swift where everything can be *coerced* to AnyObject. `protocol Foo: AnyObject` doesn't imply "Foo conforms to AnyObject", for example, because "Foo" is really `Any<Foo>`.

hborla commented:
@jrose-apple I agree, and we could definitely have a tailored diagnostic. Do you have any ideas on what we should say instead? At first I thought something like `X requires that Y be a class type`, but that sounds strange to me

hborla commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1740,6 +1740,8 @@ NOTE(where_requirement_failure_both_subst,none,
 NOTE(requirement_implied_by_conditional_conformance,none,
      "requirement from conditional conformance of %0 to %1", (Type, Type))
+NOTE(wrapped_type_satisfies_requirement,none,
+     "wrapped type %0 satisfies this requirement", (Type))
```
I agree that it would be better to say something actionable, but I can't think of anything better than a vague `; did you mean to unwrap?` or similar...any ideas??

xedin replied:
Sure, `; did you mean to unwrap?` sounds good to me!

hborla commented:
@jrose-apple I just realized that we already have a special diagnostic for this that says exactly that: `%0 requires that %1 be a class type`. It just wasn't being used in the new diagnostic

hborla committed:
[Diagnostics] Improve the error message for when a type fails to satisfy
`AnyObject` layout requirement.

jrose-apple commented:
The 100% correct thing is "%0 requires that %1 have a single reference-counted pointer representation" but that's like three separate concepts that aren't supposed to be exposed to the user…

- AnyObject is a valid `T: AnyObject`
- class instances satisfy `T: AnyObject`
- `@objc` protocol instances satisfy `T: AnyObject`

\*sigh\*

The current error message is probably good for now.

hborla committed:
[Diagnostics] For now, don't diagnose layout requirement failures
in MissingConformanceFailure::diagnoseTypeCannotConform.

hborla commented:
@swift-ci please smoke test

hborla committed:
[CSDiag] Remove diagnoseTypeRequirementFailure and
diagnoseUnresolvedDotExprTypeRequirementFailure.

hborla commented:
@swift-ci please smoke test

hborla committed:
[ConstraintSystem] Don't fail in matchTypes when two primary archetype types
are not equal when part of a type requirement. This allows the
SkipSameTypeRequirement constraint fix to be applied instead.

hborla commented:
@swift-ci please smoke test
